### PR TITLE
terraform: add completion

### PR DESCRIPTION
**Pull request by lucasew**
###### Description of changes
Terraform has completions support but  the way it's applied, by default, is through dotfile mutation. This pull request generates the completion files to allow completions without the mutation.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 223926` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>coder</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>terraform</li>
    <li>terraform_plugins_test</li>
  </ul>
</details>

File: pkgs/applications/networking/cluster/terraform/default.nix
Comment (superherointj) [inline pkgs/applications/networking/cluster/terraform/default.nix:13]:
```suggestion
```
Where do you use this?

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 223926` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>coder</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>terraform</li>
    <li>terraform_plugins_test</li>
  </ul>
</details>

File: pkgs/development/tools/coder/default.nix
Comment (superherointj) [inline pkgs/development/tools/coder/default.nix:87]:
```suggestion
```

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 223926` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>coder</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>terraform</li>
    <li>terraform_plugins_test</li>
  </ul>
</details>

**Comment by lucasew** [comment]
Result of `nixpkgs-review pr 223926` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>coder</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>terraform</li>
    <li>terraform_plugins_test</li>
  </ul>
</details>

**Review by superherointj — APPROVED** [review]

**Comment by zowoq** [comment]
@superherointj Please let the package maintainers review PRs. There was any urgency to merge this and you aren't the maintainer.

File: pkgs/applications/networking/cluster/terraform/default.nix
Comment (zowoq) [inline pkgs/applications/networking/cluster/terraform/default.nix:59]: Why was this reformatted?

Comment (superherointj) [reply pkgs/applications/networking/cluster/terraform/default.nix:59]: To remove double `with` at meta.
Reasoning: https://github.com/atorres1985-contrib/nixpkgs/commit/451dfa6d84586b5cb83e6a7254689dff66b4523a

Comment (zowoq) [reply pkgs/applications/networking/cluster/terraform/default.nix:59]: That isn't nixpkgs policy.

Comment (superherointj) [reply pkgs/applications/networking/cluster/terraform/default.nix:59]: > That isn't nixpkgs policy.

And what is the policy?

**Comment by zowoq** [comment]
Terraform completion doesn't work for me on macos with zsh.

File: pkgs/applications/networking/cluster/terraform/default.nix
Comment (superherointj) [reply pkgs/applications/networking/cluster/terraform/default.nix:59]: Feel free to revert whatever you want.

**Comment by superherointj** [comment]
> Terraform completion doesn't work for me on macos with zsh.

Feel free to fix it.

**Comment by zowoq** [comment]
I'm reverting it. https://github.com/NixOS/nixpkgs/pull/223973

**Comment by lucasew** [comment]
> Terraform completion doesn't work for me on macos with zsh.

If you do like `nix-shell -p terraform` it doesn't work

But in the nixpkgs-review shell it works nice.

**Comment by zowoq** [comment]
I didn't test it with nix-shell, I switched my profile to master after this was merged.

I confirmed that the completion file was installed, it didn't work.

File: pkgs/applications/networking/cluster/terraform/default.nix
Comment (lucasew) [reply pkgs/applications/networking/cluster/terraform/default.nix:59]: That's to remove the `with`. That was @superherointj idea, though.

**Comment by lucasew** [comment]
> I didn't test it with nix-shell, I switched my profile to master after this was merged.
> 
> I confirmed that the completion file was installed, it didn't work.

Weird

But stufff seems to be in the right place.

File: pkgs/applications/networking/cluster/terraform/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/networking/cluster/terraform/default.nix:74]: It does not work because you mixed the vars up here.

**Comment by SuperSandro2000** [comment]
> I'm reverting it. #223973

Instead of reverting things right away, we should take a minute or two to see if we can fix things. It took me not even a minute to spot the mistake and I am sure you would have spotten it, too if you would have just looked at the change instead of fixating on the revert straight away.

**Comment by zimbatm** [comment]
The maintainer should get to decide at what pace they want to work. If somebody merges something, without giving the opportunity to the maintainer to review it, and doubly so if the change is broken, then the best is to revert it, take a step back, and then re-try. It's not fair to ask the maintainer to rush to work on a fix that they didn't approve.

File: pkgs/applications/networking/cluster/terraform/default.nix
Comment (zowoq) [reply pkgs/applications/networking/cluster/terraform/default.nix:74]: Can you be more specific about what you think is incorrect here please?

**Comment by SuperSandro2000** [comment]
Yeah, I thought the fix was pretty easy but apperantly I need to find my glasses. Sorry about that.

**Pull request merged**
